### ISSUE TITLE
Agents seem to have fewer cells in range than their vision should let them see (including disease effects)

### ISSUE THREAD

**Issue by nkremerh** [nkremerh/sugarscape#56]
After digging into `ethics.py` to change out some hardcoded values, I noticed that agents are frequently having fewer cells found in the `cellsInRange` attribute (which should be equivalent to the number of cells found by a call to `findVision()` multiplied by the kind of neighborhood).

Using the default von Neumann neighborhood scheme, an agent should have `agent.findVision() * 4` cells available to consider. There are cases where a healthy agent (so no disease effects) is seeing `(agent.findVision() - 1) * 4` cells. This is not universal, which is puzzling. This led me to check for disease, but healthy agents are experiencing this behavior. It should also be noted that not _all_ healthy agents experience this, either.

This is apparent in seed `12345` using the rest of the default configuration options (recommended to ratchet down the timesteps to 200 or fewer and set the decision model to `benthamHalfLookaheadBinary`).

**Comment by colinhanrahan** [comment]
Although the agents are healthy, when `startingDiseases` is 0, this behavior never occurs. At a surface glance, it appears like they're not recovering from diseases correctly. 

Adding `self.findCellsInRange()` when curing diseases in `doDisease()` resolves all occurrences of the bug I can find:
```
if diseaseTags == immuneResponseCheck:
    self.diseases.remove(diseaseRecord)
    self.updateDiseaseEffects(diseaseRecord["disease"])
>>> self.findCellsInRange()
```
On the timestep a disease is cured,
- `cellsInRange` is calculated in the call stack of `moveToBestCell()` (when the agent still has a disease).
- Then, the disease is cured later in the timestep at `doDisease()`.

If you’re checking expected vs. actual cells in range at the _end_ of each timestep, it technically won’t line up. I don't think this affects the simulation because nothing related to `cellsInRange` happens after `doDisease()`, and it will be rectified on the next timestep anyway. You're welcome to add the change if it makes the program more intuitive—it should have minimal effect on runtime.

Let me know if you're still seeing this bug occur with the fix included.

**Comment by colinhanrahan** [comment]
You're right—the discrepancy occurs in `canReachCell()`. Agents later in the timestep use their neighbors’ `cellsInRange` (the ones who possibly just got cured) to calculate `certainty`, which is a big factor in the ethical value of the cell. `certainty = 0` when `neighbor.canReachCell(cell) == False` cancels out the entire calculation for that neighbor. Setting the decision model to `none` where `canReachCell()` is never called seems to maintain determinism with the current repo head. I do recommend pushing this now because it makes sense for agents to have the correct `cellsInRange` when other agents are assessing them.

**Comment by colinhanrahan** [comment]
Actually, `doDisease()` isn't the best place to update `findCellsInRange()`. It should be updated _every_ time an agent moves, even a healthy one, because otherwise the information will be incorrect for calculating `certainty`. Placing it just after the death check in `doTimestep()` should do the trick. (This will also help with another issue I'm pursuing.)
```python
    # If dead from aging, skip remainder of timestep
    if self.alive == False:
       return
+   self.findCellsInRange()
    self.updateHappiness()
```
For efficiency, it should then be removed (for default behavior) from `findNeighborhood()`:
```python
    def findNeighborhood(self, newCell=None):
-      newNeighborhood = self.findCellsInRange(newCell)
+      if newCell == None:
+          newNeighborhood = self.cellsInRange
+      else:
+          newNeighborhood = self.findCellsInRange(newCell)
```

**Comment by nkremerh** [comment]
@colinhanrahan you are planning on one final PR to address the rest of this issue, correct? Or is this ready to close with #66?

**Comment by colinhanrahan** [comment]
@nkremerh Well, it turns out that finding the Euclidean distance to a cell is about as fast as checking every cell in its range (most of the time). Still, checking distance rather than `cellsInRange` for `canReachCell()` would allow us to separate agent timesteps more than they already are, which is good—overlap between timesteps can be tricky. (I left a note on the above PR explaining another change we need to make regarding this.) I still don't know why the population is lower, so I'm a little wary of the solution we have now.

In terms of speed, the fastest solution is to use a set for `cellsInRange` for the quick membership check. However, dictionaries aren't hashable, so we'd need to use a tuple or something similar.

To answer your question, there's one small change we need to make (again, see above PR) and then we need to decide on the best solution for `canReachCell`—current repo vs. distance check vs. making `cellsInRange` a set.

**Comment by nkremerh** [comment]
Resolved by #66 and #78.
